### DISCUSSION TITLE
New structure to level 7 and up

### DISCUSSION THREAD

**Discussion by MarleenGilsing** [hedyorg/hedy#903] [Language Ideas]
I was talking with @Felienne and we came up with this possible new structure of the levels 7 - 9. 

Level 6 could stay as is. Maths but only with whole numbers.  
Level 7: Introducing blocks and indention. Very similiar to how it is now, but we'd make it impossible to put blocks inside other blocks. So repeat and if can be used on multiple lines of code, but they can't be nested.
Level 8: Allow for nesting if and repeat commands. 
Level 9: 'Kommagetallen and friends' as Felienne put it. This level will be dedicated to teaching the students different types of data (e.g. text vs numbers). This level will allow decimal numbers and should teach them to put quotation marks around texts, and allow students to use the < and > signs. 

@OnnoEbbens Of course we'd like your opinion on this topic too.

**Comment by OnnoEbbens** [comment]
I really like the idea to introduce data types in level 9, I think that is the perfect moment for that. 

I am not a big fan of splitting level 7 into two levels. I introduced level 7 last week and one of the first questions was if we can nest loops and if-statements. I am a bit afraid that the change from level 7 to level 8 will become too small. When students get to level 7 they are familiar with Hedy and their pace increases significantly. So I think they can cope with bigger steps. However we are still in the middle of level 7 so I might be wrong and overestimate my students.

**Reply by Felienne** [reply]
> I really like the idea to introduce data types in level 9, I think that is the perfect moment for that.
> 
> I am not a big fan of splitting level 7 into two levels. I introduced level 7 last week and one of the first questions was if we can nest loops and if-statements. I am a bit afraid that the change from level 7 to level 8 will become too small. When students get to level 7 they are familiar with Hedy and their pace increases significantly. So I think they can cope with bigger steps. However we are still in the middle of level 7 so I might be wrong and overestimate my students.

Hi Onno! I am curious if you have updates on this. In the dataset we do see a lot of issues with nesting in nesting so I think it is  good choice. I made a special new error message that warns that the kids is doing well but that block in block is not allowed yet ("block in block" is not yet translated but it will be)
![image](https://user-images.githubusercontent.com/1003685/139384143-6e05e843-9c71-47f1-a976-53d9c31deb3f.png)

**Comment by MarleenGilsing** [comment]
As addition to this discussion, @Felienne and I also discussed the new purpose of lvl 9 - 11. 
We are now planning on structuring the Hedy levels like this: 

Lvl 5 - maths
Lvl 6 - repeat 
Lvl 7 - if with multiple lines of code
Lvl 8 - nesting repeat and if 
Lvl 9 (new)  - for animal in animals
Lvl 10 (new)  - for i in range 1 to 10
Lvl 11 (new)  - datatypes

**Comment by OnnoEbbens** [comment]
Hi! I don't have much news on this, right after I wrote my comment we switched to Python. I think the new structure with an extra level for nested if-statements and for-loops is good. Also I like the new start of the for-loop with a list instead of a range.

**Comment by Felienne** [comment]
Hi!

Continuing the discussion further about higher levels here. We currently have this implemented up to 10 and that will be merged soon (via #1144):

- Lvl 5 - maths
- Lvl 6 - repeat
- Lvl 7 - if with multiple lines of code
- Lvl 8 - nesting repeat and if
- Lvl 9 (new) - for animal in animals
- Lvl 10 (new) - for i in range 1 to 10

Still to be discussed and planned are:

**Lvl 11 (new) - datatypes** 

In this level we explain that 5 is different from '5' and form 5.0. We also show that some operations like '+' have different behaviour on text than on numbers. Allows for, for example, more complex pocket money calculations and ASCII are like visualisations  with "-" * number.

**Lvl 12 (new) - Bigger and smaller (now level 17)**

Learners learn about < and > in preparation for while loops. I think we can also merge in <= and >= and maybe even != that are now in level 21 and 22 after the equal sign (What do you think @MarleenGilsing? Is that too much in one step?)

**Lvl 13 (new) - While loops (now level 18)**

Learners are introduced to the while loop. With the previous knowledge of booleans and < and >, learners
can make basic while loops.

**Lvl 14 (new) - Access list values (starting at 1) (now level 19)** 

Learners are shown how to get a specific value from a list, like `names[1]`. This couples nicely with while loops, and allows for various new types of programs, for example summing all numbers in a list (nice continuation of new level 11) or making a leaderboard of scores.

Stuff we then still need to teach after this:
* Lists starting at 0 instead of 1 (hard mental shift)
* True and False (tricky but could be handy for the while loop?)
* elif in if-else (was once level 10) allows for nicer programs but does not really add much more power
* is becomes = and == (hard *and* boring)
* Round brackets and square brackets (now level 11 and 12) (somewhat hard and boring)
* Colons after if and for and while (easy but boring)

None of these are really concepts but they are needed for a switch to Python

Stuff we could teach (before adding some syntax like colons):
* functions
* debugging (we thought of that as a separate level before)
* dictionaries
* Anything else?

**Reply by MarleenGilsing** [reply]
@Felienne summary of what we just discussed: 
is should change to = in level 5. 
== should be added to level 12, together with > >= and != etc

**Reply by MarleenGilsing** [reply]
@Felienne Where did you plan to introduce elif ( maybe together with the : after for/if/while, like the current level 10).

**Reply by Felienne** [reply]
> @Felienne Where did you plan to introduce elif ( maybe together with the : after for/if/while, like the current level 10).

What do you think? I think it is nice to mix it in with another feature, because in itself it does not add all that much, but maybe not the : because that is syntax is this is more a concept.

We could also just put it in when we introduce multiline if so kids can use it when they need it? Of course that leaves the question of when and how to teach it.

**Reply by Felienne** [reply]
> **Lvl 14 (new) - Access list values (starting at 1) (now level 19)**
> 
> Learners are shown how to get a specific value from a list, like `names[1]`. This couples nicely with while loops, and allows for various new types of programs, for example summing all numbers in a list (nice continuation of new level 11) or making a leaderboard of scores.

Ok I was working on this a bit today but I start to doubt now if this really adds nice programs, so I a need a bit of brain power to help me think about that!

The problem is that contrary to my previous message, there are not all that much things you need indexing for, except for kinda boring things like sorting a list. We could invent a few games like "guess at what place in the list an X is stored but that is kind of contrived" Things you would really need indexing for that kids know need 2D grids and not lists, like tic-tac-toe or minesweeper. We can of course teach those or plot flat lists as 2d but that is all so "vergezocht".

I think the best is to merge this in with also adding [] for list definitions and add and remove, then we add some nice features related to lists as one package. This is sort of what we did with < and <= also, so a general thing that we are seeing, apparently, is that higher levels may have larger steps with a few features grouped together, so we can really follow through on the promise of relevant things to do in each level.
